shards: implement markReady on shardedSearcher

keegancsmith opened the pull request:
This is the final commit implementing a non-zero Crashes value while
shardedSearcher is startin up. This is meant to communicate when a zoekt
is partially available. When landed, Sourcegraph's APIs will communicate
that the user may have partial results and should retry.

Test Plan: go test. Additionally added a large sleep to loader.load to
simulate a slow startup. Then did searches in zoekt-webserver and
observed it returning non-zero crash counts until all shards loaded.

plz-review-url: https://plz.review/review/15657

stefanhengl reviewed (approved):
If I understand correctly, this is just for the initial load. Loading shards that were just indexed will not be reported as partial availability, correct?

plz-review[bot] committed:
shards: implement markReady on shardedSearcher

This is the final commit implementing a non-zero Crashes value while
shardedSearcher is startin up. This is meant to communicate when a zoekt
is partially available. When landed, Sourcegraph's APIs will communicate
that the user may have partial results and should retry.

Test Plan: go test. Additionally added a large sleep to loader.load to
simulate a slow startup. Then did searches in zoekt-webserver and
observed it returning non-zero crash counts until all shards loaded.

plz-review-url: https://plz.review/review/15657

keegancsmith commented:
> If I understand correctly, this is just for the initial load.

Yes.

> Loading shards that were just indexed will not be reported as partial availability, correct?

Yes, follow-up loads will not suddenly report us as partial availability. Mainly because when we update later on we are normally replacing and we do it in a way were we never stop serving a specific repo.
